services/apps/data_sink_worker/src/service/member.service.ts

```diff
@@ -250,13 +255,15 @@ export default class MemberService extends LoggerBase {
     for (const email of emails) {
       if (email) {
         const domain = email.split('@')[1]
-        emailDomains.add(domain)
+        if (!isDomainExcluded(domain)) {
+          emailDomains.add(domain)
+        }
       }
     }
 
     // Assign member to organization based on email domain
     for (const domain of emailDomains) {
-      const org = await orgService.findByDomain(tenantId, segmentId, domain as string)
+      const org = await orgService.findOrCreateByDomain(tenantId, segmentId, domain as string)
```

Comment:
isn't it better to only find the existing organization here, if it exists send that organization, and if it doesn't prepare the payload for below. We already loop through this organization array and call `findOrCreate` anyway - This way it's called twice - once here (`findOrCreateByDomain`) and once when we loop through organizations

also if it doesn't exist yet and we're gonna create it, we need to append an identities array to the found organization by domain (We can set the platform as e-mail and identity as domain name found)
it can be sent in `organization.identities` in the form
```
identities: [
     {
         platform: 'email',
         name: 'someDomain',
     }
]
```

Also I think we should check if the organization already exists in the sent `member.organizations` array to prevent multiple calls to `findOrCreate` for the same organization

what do u think?

Reply:
Thanks for reviewing! regarding the code, I don’t think `findOrCreate` is called twice, that's because both work on different scopes and we don’t loop over the organization array but rather data.organizations

If we try to find only by domain and return if there is org else payload to create one, we might still end up running the find query more than once (in `findByDomain` and `findOrCreate`) Also, that makes it quite tricky to pass it as argument to `findOrCreate` method.

To clarify the question in short: `findOrCreate` is called only if the data.organizations is true and not loop through organizations list.

Let me know, if you think we could still improve it 🙂